packages/common-utils/src/core/renderChartConfig.ts

```diff
@@ -281,6 +292,250 @@ const fastifySQL = ({
+    numNode = node.left as SQLParser.Value;
+  } else {
+    return nope;
+  }
+
+  // check numNode
+  if (numNode.type !== 'number' || numNode.value !== 1) return nope;
+
+  // check funcNode to be "notEmpty(map, key)"
+  if (funcNode.name.name[0]?.value !== 'notEmpty') return nope;
+  if (funcNode.args?.value?.length !== 1) return nope;
+  const argsNode = funcNode.args.value[0];
+  if (argsNode.type !== 'column_ref') return nope;
+  const ident = extractIdent(argsNode as ColumnRef);
+  if (!ident) return nope;
+
+  return {
+    isMapValueNullCondition: true,
+    ident,
+  };
+};
+
+const optimizeMapAccessWhere = ({
+  columns,
+  rawSQL,
+}: {
+  columns: ColumnMeta[];
+  rawSQL: string;
+}): string => {
+  // Parse the SQL AST
+  const idents: Array<{
+    ident: SQLValueIdent;
+    doesContain: boolean;
+  }> = [];
+  try {
+    const parser = new SQLParser.Parser();
+    const ast = parser.astify(rawSQL, {
+      database: 'Postgresql',
+    }) as SQLParser.Select;
+
+    // traverse ast and collect all column or map identifiers that are used
+    const traverse = (
+      node:
+        | SQLParser.Expr
+        | SQLParser.ExpressionValue
+        | SQLParser.ExprList
+        | SQLParser.Function
+        | null,
+    ) => {
+      if (node == null) {
+        return;
+      }
+
+      const traverseNode = (
+        nodes: SQLParser.ExpressionValue | SQLParser.ExprList,
+      ) => {
+        if (Array.isArray(nodes)) {
+          for (const node of nodes) {
+            traverse(node);
+          }
+        } else {
+          traverse(nodes);
+        }
+      };
+
+      switch (node.type) {
+        case 'binary_expr': {
+          // Could be =, !=, AND, OR, LIKE, ILIKE
+          const _n = node as SQLParser.Expr;
+
+          if (_n.operator === '!=') {
+            const { isMapValueNullCondition, ident } = checkIsMapValueNull(_n);
+            if (isMapValueNullCondition && ident) {
+              idents.push({
+                ident,
+                doesContain: false,
+              });
+              // do not traverse
+              break;
+            }
+          }
+          traverseNode(_n.left);
+          traverseNode(_n.right);
+          break;
+        }
+        case 'function': {
+          const _n = node as SQLParser.Function;
+          if (_n.args?.type === 'expr_list') {
+            traverseNode(_n.args?.value as any);
+          }
+
+          break;
+        }
+        case 'column_ref': {
+          const ident = extractIdent(node as ColumnRef);
+          if (ident) {
+            idents.push({ doesContain: true, ident });
+          }
+          break;
+        }
+        default:
+          // ignore other types
+          break;
+      }
+
+      return;
+    };
+
+    traverse(ast.where);
+
+    if (idents.length === 0) {
+      return parser.sqlify(ast);
+    }
+
+    // build set of Maps
+    const maps = new Set(
+      columns.filter(v => v.type.startsWith('Map')).map(v => v.name),
+    );
```

Comment:
perf: We can move this out of the method, and we don’t need to traverse the AST if it’s empty. Also `maps` is a bit ambiguous, maybe `mapFieldNames` ?